fix(session-replay): improve multi-threading of session replay processing

philprime opened the pull request:
## :scroll: Description

- Changes the priorities of the background queues of session replay.
- Refactors the async dispatch handling of session replay.

## :bulb: Motivation and Context

Some of the methods of the session replay video processing are dispatched to a low priority background queue, which then dispatches additional AV work on another dispatch queue.

The current implementation does not respect queue priorities and can cause thread starvation, because the processing queue and the work queue are either on the same priority, or the work queue is on a lower priority.

In the worst case, the processing queue is blocked and waits for the work queue, but the work queue has a lower priority than the processing queue and is awaiting execution on the main thread.

This is what the thread inversion warning is showing.

Closes #4730

## :green_heart: How did you test it?

Manual testing.

## :pencil: Checklist

You have to check all boxes before merging:

- [ ] I added tests to verify the changes.
- [X] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [X] I updated the docs if needed.
- [X] I updated the wizard if needed.
- [X] Review from the native team if needed.
- [X] No breaking change or entry added to the changelog.
- [X] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

philprime commented on Sources/Sentry/SentrySessionReplayIntegration.m:
```diff
@@ -189,30 +205,54 @@ - (void)resumePreviousSessionReplay:(SentryEvent *)event
+    __block NSArray<SentryVideoInfo *> *videos;
+    __block NSError *_Nullable error;
+
+    dispatch_group_t group = dispatch_group_create();
```
I would appreciate feedback if you asee any issues with this approach for synchronizing the async call

philprime commented on Sources/Sentry/SentrySessionReplayIntegration.m:
```diff
@@ -316,30 +358,27 @@ - (void)startWithOptions:(SentryReplayOptions *)replayOptions
         = (NSInteger)(shouldReplayFullSession ? replayOptions.sessionSegmentDuration + 1
                                               : replayOptions.errorReplayDuration + 1);
 
-    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(
```
Moved this setup to the SentryDispatchQueueWrapper to make it replaceable

philprime commented:
This PR grew into a a rather large refactor, I'll create additional PRs to split the changes into smaller digestible parts

philprime commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -120,128 +130,260 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+        do {
+            videoWriter = try AVAssetWriter(url: outputFileURL, fileType: .mp4)
+        } catch {
+            SentryLog.debug("[Session Replay] Failed to create video writer, reason: \(error)")
```
```suggestion
            SentryLog.error("[Session Replay] Failed to create video writer, reason: \(error)")
```

philprime commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -120,128 +130,260 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+    private func renderVideo(with videoFrames: [SentryReplayFrame], from: Int, at outputFileURL: URL, completion: @escaping (Result<SentryRenderVideoResult, Error>) -> Void) {
+        SentryLog.debug("[Session Replay] Rendering video with \(videoFrames.count) frames, from index: \(from), to output url: \(outputFileURL)")
+        guard from < videoFrames.count, let image = UIImage(contentsOfFile: videoFrames[from].imagePath) else {
+            SentryLog.debug("[Session Replay] Failed to render video, reason: index out of bounds or can't read image at path: \(videoFrames[from].imagePath)")
```
```suggestion
            SentryLog.error("[Session Replay] Failed to render video, reason: index out of bounds or can't read image at path: \(videoFrames[from].imagePath)")
```

philprime commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -120,128 +130,260 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+
+    func createVideoAsyncWith(beginning: Date, end: Date, completion: @escaping ([SentryVideoInfo]?, Error?) -> Void) {
+        // Note: In Swift it is best practice to use `Result<Value, Error>` instead of `(Value?, Error?)`
+        //       Due to interoperability with Objective-C and @objc, we can not use Result here.
```
```suggestion
        //       Due to interoperability with Objective-C and @objc, we can not use Result for the completion callback.
```

philipphofmann commented:
I'm unsure if we're choosing the right approach here by adding two more DispatchQueues. If I'm not mistaken, the root cause is using the default DispatchQueue of the `SentryDispatchQueueWrapper` here: 
https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift#L169-L170

`SentrySessionReplayIntegration.resumePreviousSessionReplay` already runs on a BG thread scheduled on the `SentryDispatchQueueWrapper.queue`. This `resumePreviousSessionReplay` then calls `SentryOnDemandReplay.createVideoWithBeginning`, which calls `SentryOnDemandReplay.renderVideo`. `renderVideo` then renders the video with `AVAssetWriterInput.requestMediaDataWhenReady`, which uses `SentryDispatchQueueWrapper.queue`. As this queue is serial, I'm not surprised that, especially when the SDK starts and there are a couple of envelopes that need to be sent, we receive a priority inversion warning. 

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift#L147-L206

I don't think we need to call `createVideoAsyncWithBeginning` in the `SentrySessionReplayIntegration`. Creating the video already happens on a BG thread. Making it async requires using a DispatchGroup below, so it's more complex again.

https://github.com/getsentry/sentry-cocoa/blob/44c4af72f68c588083c5c351b853ed8b76821275/Sources/Sentry/SentrySessionReplayIntegration.m#L231-L244

Instead, I think we need to use one extra dedicated queue for video processing in `SentryOnDemandReplay` and this should solve the issue. Maybe I'm missing something. I'm also happy to brainstorm this on a call.

philprime commented:
I'll try to go into your point, but yes I'll contact you regarding a call to discuss this in detail.


> If I'm not mistaken, the root cause is using the default DispatchQueue of the SentryDispatchQueueWrapper here:

Yes and no. The actual problem causing the warning is this line...

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift#L202

... as it was called from a queue which was created here:

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Sentry/SentrySessionReplayIntegration.m#L333-L337

This serial queue with priority low was still somehow on a higher priority than the one used by the asset writer in `requestMediaDataWhenReady`, while being unclear on which priority. 

This `DISPATCH_QUEUE_PRIORITY_LOW` seems to not even exist in Swift [`DispatchQueue.QoSClass`](https://developer.apple.com/documentation/dispatch/dispatchqos/qosclass-swift.enum) which is why I changed it to `QOS_CLASS_BACKGROUND` here:

https://github.com/getsentry/sentry-cocoa/blob/44c4af72f68c588083c5c351b853ed8b76821275/Sources/Sentry/SentryDispatchQueueWrapper.m#L112-L113

This change alone might already fix the Xcode warning, but this PR is in general to improve multi-threading to ensure that thread inversion can not happen.

> I'm unsure if we're choosing the right approach here by adding two more DispatchQueues. 

> I don't think we need to call `createVideoAsyncWithBeginning` in the `SentrySessionReplayIntegration`. Creating the video already happens on a BG thread. Making it async requires using a DispatchGroup below, so it's more complex again.

We are not adding two more queues, we are only adding the `asset worker queue`.

The `processing queue` already existed here:

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Sentry/SentrySessionReplayIntegration.m#L333-L337

But instead of using it in the `SentrySessionReplay.swift` here (only use case in the class):

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Swift/Integrations/SessionReplay/SentrySessionReplay.swift#L226-L242

I moved the background dispatching into the `createVideoWith(beginning:)` so that the class `SentrySessionReplay` **does not do any dispatching anymore**, but instead uses a callback approach to process the video segments ([PR diff](https://github.com/getsentry/sentry-cocoa/pull/5018/files#diff-7e90bc4012112a489d229dd525a4bb5c21ee3478152aae0f1c0b75b34aa85947R221-R239)).

The processing queue is therefore now moved to the `SentryOnDemandReplay`, but instead of just creating it in the initializer, I made it injectable from `SentrySessionReplayIntegration.m`:

https://github.com/getsentry/sentry-cocoa/blob/44c4af72f68c588083c5c351b853ed8b76821275/Sources/Sentry/SentrySessionReplayIntegration.m#L134-L139

> Instead, I think we need to use one extra dedicated queue for video processing in `SentryOnDemandReplay` and this should solve the issue. Maybe I'm missing something. I'm also happy to brainstorm this on a call.

That's what the newly added `asset worker queue` is for:

https://github.com/getsentry/sentry-cocoa/blob/44c4af72f68c588083c5c351b853ed8b76821275/Sources/Sentry/SentrySessionReplayIntegration.m#L129-L133

As mentioned above, the `createVideoAsyncWith(beginning:)` is dispatching into the background processing queue to iterate all the frames and convert them into segments...

https://github.com/getsentry/sentry-cocoa/blob/44c4af72f68c588083c5c351b853ed8b76821275/Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift#L140C10-L147

... and we need to loop through multiple frames and segments, while waiting for the `renderVideo` callback to respond, therefore it uses a semaphore to synchronize it.

The reason why `renderVideo` is async, is caused by the AVAssetWriter using some self-configured background queue inside which is outside our control (as I did not see any way of configuring it), therefore instead we can only pass a callback queue to `requestMediaDataWhenReady`. Now as the `processing queue` is already waiting using the semaphore, the completion block of `requestMediaDataWhenReady` can not be dispatched on this blocked queue, and instead we need the `asset worker queue` to process media, then call the callback of `renderVideo` to then releasing the semaphore and the dispatch queue.

If we go back to the previous implementation of `renderVideo` which was not async with a callback but synchronized itself using DispatchGroup...

https://github.com/getsentry/sentry-cocoa/blob/38fc771279e2920cdb07f11c9cbe12ce4836a3e7/Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift#L164-L171

... that would be the same logic as if we do the synchronization 
e could go back to the previous implementation of the `renderVideo` which did not offer a callback but synchronized itself using dispatch group, but that would still be the same as the logic in `createVideoAsyncWith`.

This is therefore also a shift of synchronization logic, but I don't understand why `renderVideo` should be synchronized in the first place, if the used `viewWriterInput` is also using callback-style async.

Finally the `asset worker queue` has an explicitly defined higher priority than the `processing queue`, so when the `processing queue` waits for the `asset worker queue` the dispatcher will prefer the `asset worker` to finish and release the other one.

philprime commented:
I found [this PR comment](https://github.com/getsentry/sentry-cocoa/pull/4194#discussion_r1700050798) and #4439 which I'll need to go through in detail again because the comments do include additional information on the choices made.

I'll try to summarize the reasonings and add more comments to the code itself, so it is documented for future maintenance.

philprime commented:
I also added more logging where relevant to debug issues

philipphofmann commented on CHANGELOG.md:
```diff
@@ -18,6 +18,10 @@
 - Correctly rate limit envelopes from the new UI profiling system (#5131)
 - Race condition in ANRTrackerV1 (#5137)
 
+### Fixes
+
+- Fix thread inversion warning in session replay (#5018)
```
`m`: There is already a fixes section for the unreleased section

philipphofmann commented on Sources/Sentry/SentrySessionReplayIntegration.m:
```diff
@@ -121,6 +123,19 @@ - (void)setupWith:(SentryReplayOptions *)replayOptions
+
+    // The asset worker queue is used to work on video and frames data.
+    // Use a relative priority of -1 to make it lower than the default background priority.
+    _replayAssetWorkerQueue = [SentryDispatchQueueWrapper
+        createBackgroundDispatchQueueWithName:"io.sentry.session-replay.asset-worker"
+                             relativePriority:-1];
+    // The dispatch queue is used to asynchronously wait for the asset worker queue to finish its
+    // work. To avoid a deadlock, the priority of the processing queue must be lower than the asset
+    // worker queue. Use a relative priority of -2 to make it lower than the asset worker queue.
+    _replayProcessingQueue = [SentryDispatchQueueWrapper
+        createBackgroundDispatchQueueWithName:"io.sentry.session-replay.processing"
```
`h`: The comment mentions that setting the wrong priorities here can lead to a deadlock. I think it's essential to have a test fail when using the wrong priorities to avoid regressions. We could assert that the `SentryDispatchQueueWrapper` is initialized with the correct attributes, which tests implementation details, and has the downside of having to change the test when the implementation details change. We could also have a test causing a deadlock when interacting with this class when the priorities are wrongly set, which could be a bit harder to achieve, but has the advantage of not testing implementation details.

philprime replied:
See my answer to the other comments on how I'll make sure that SentryDispatchQueueWrapper is called correctly.

I am not sure yet if it's viable to create a failing test if the priorities are wrong, as we are doing this because the inner workings of AVFoundation are not transparent, but Xcode shows a warning.

philprime replied:
As mentioned in the other PR comment, I added tests

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -122,150 +141,257 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+            // It is imporant that the renderVideo completion block signals the group.
+            // The queue used by render video must have a higher priority than the processing queue to reduce thread inversion.
+            // Otherwise, it could lead to queue starvation and a deadlock/timeout.
+            guard group.wait(timeout: .now() + 120) == .success else {
```
`m`: 120 seems a bit high to me. Shouldn't 30 be enough?

philprime replied:
Thanks for pointing this out! 🙏 
I had to set this higher because it was set to 2 seconds, and when debugging the timeout would always hit, but forgot to revert it afterwards.

philipphofmann commented on Sources/Sentry/SentrySessionReplayIntegration.m:
```diff
@@ -121,6 +123,19 @@ - (void)setupWith:(SentryReplayOptions *)replayOptions
     _notificationCenter = SentryDependencyContainer.sharedInstance.notificationCenterWrapper;
+    _dateProvider = SentryDependencyContainer.sharedInstance.dateProvider;
+
+    // The asset worker queue is used to work on video and frames data.
+    // Use a relative priority of -1 to make it lower than the default background priority.
+    _replayAssetWorkerQueue = [SentryDispatchQueueWrapper
+        createBackgroundDispatchQueueWithName:"io.sentry.session-replay.asset-worker"
+                             relativePriority:-1];
```
`h`: That's so easy to break. If we ever change the priority of the default background priority here, we could end up with the same problem again. And when I change the priority in the `SentryDispatchQueueWrapper` how should I know that I I need to change it here too?

https://github.com/getsentry/sentry-cocoa/blob/8856e5a7cf210e0417952d909d5924e76a7d1620/Sources/Sentry/SentryDispatchQueueWrapper.m#L10-L13

A test should fail if these priorities don't match our requirements here. If you need ideas for testing this properly, please let me know.

philprime replied:
I would highly appreciate an example how you would test the priorities of queues.

philipphofmann replied:
Using a factory method that I can swap out in the tests or something like that. Then I can assert that I call the factory with the correct parameters.

philprime replied:
Thinking about this again, I am starting to question if another factory makes sense, because the `createBackgroundDispatchQueueWithName` already is a factory.
But I do see your point of this breaking easily and that we do not have tests to validate the behavior.

Therefore I am going to implement this:
- I'll add a new protocol `SentryDispatchQueueProviderProtocol` and an instance of `SentryDispatchQueueProvider`
- The provider will then use the `SentryDispatchQueueWrapper` under the hood to create dispatch queues
- This will make it replaceable in tests so we can assert that the queues are created with the correct arguments.
- The provider will live in the container, as there is no possibility to inject a dependency here, as we are already in the top-level integration

philprime replied:
Nevermind, I just noticed that the `SentryDispatchQueueWrapper` already is used as an instance to create queues, so I'll just change the new method to be non-static

philprime replied:
I just found `SentryDispatchFactory`, will extend this

philprime replied:
alright, added the protocol and tests for the queues.
I actually did not mock the factory in the unit tests, but instead just checked the priorities of the queues underneath. Still believe there is value in adopting the factory pattern.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -122,150 +141,257 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+            // It is imporant that the renderVideo completion block signals the group.
+            // The queue used by render video must have a higher priority than the processing queue to reduce thread inversion.
+            // Otherwise, it could lead to queue starvation and a deadlock/timeout.
+            guard group.wait(timeout: .now() + 120) == .success else {
+                SentryLog.error("[Session Replay] Timeout while waiting for video rendering to finish.")
+                throw SentryOnDemandReplayError.errorRenderingVideo
```
`m`: If we timeout, we could still return the videos we have and send what we have instead of discarding everything.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -122,150 +141,257 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+            }
+
+            // If there was an error, throw it to exit the loop.
+            if let error = currentError {
+                SentryLog.error("[Session Replay] Error while rendering video: \(error), cancelling video segment creation")
+                throw error
+            }
```
`m`: Same here, we could still return the videos we have and try to send them instead of throwing away all of them.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -122,150 +141,257 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+            // It is imporant that the renderVideo completion block signals the group.
+            // The queue used by render video must have a higher priority than the processing queue to reduce thread inversion.
+            // Otherwise, it could lead to queue starvation and a deadlock/timeout.
+            guard group.wait(timeout: .now() + 120) == .success else {
```
`m`: What drove you to add the DispatchGroup inside the `while frameCount < videoFrames.count {` and not outside? Now we have a timeout of 120 for every loop iteration, which can be a very long time. Shouldn't we have one timeout for the whole process?

philprime replied:
The timeout is wrong as mentioned in the other comment. It was at 2 seconds before, which is also a very tight window to do video conversion. I will reconsider a longer timeout for the entire process instead.

philprime replied:
I changed it to 10 seconds, reasonable amount for a video conversion

philipphofmann reviewed (commented):
This is going in the right direction. I have a few questions on error handling and testing. Apart from that it already looks way better than before.

philprime commented:
After looking at the internals, I noticed that we are propagating the error from the video creation to the caller, then just log and discard it because there's nothing else we can do.

I removed the propagation of the error and logged it directly `SentryOnDemandReplay.createVideoWith`, but instead return any previously generated videos.

This allowed me to get rid of multiple error handlers and try-throws

philipphofmann commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+### Fixes
+
+- Fix thread inversion warning in session replay (#5018)
```
`m`: Yes this fixes the warning, but what's the customer benefit? Faster video conversion and no deadlocks or such? If yes, can we rephrase this?

philprime replied:
Updated, should be good now

philipphofmann commented on Tests/SentryTests/Integrations/SessionReplay/SentrySessionReplayIntegrationTests.swift:
```diff
@@ -638,6 +638,29 @@ class SentrySessionReplayIntegrationTests: XCTestCase {
         XCTAssertEqual(writtenLastData, currentData)
     }
 
+    func testQueuePriorities_processingQueueShouldHaveLowerPriorityThanWorkerQueue() throws {
```
Thanks for the test 🥇

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift:
```diff
@@ -135,164 +146,264 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+
+            // If there was an error, log it and exit the loop.
+            if let error = currentError {
+                // In previous implementations the error was propagated to the completion block, discarding any generated video.
```
`l`: Mentioning a specific SDK version makes this a bit more specific, because if I want to find out which one, I need to do some git blaming.

philipphofmann reviewed (approved):
Thanks, LGTM
